torch/lib/THD/master_worker/master/generic/THDTensorMath.cpp

```diff
@@ -0,0 +1,314 @@
+#ifndef TH_GENERIC_FILE
+#define TH_GENERIC_FILE "master_worker/master/generic/THDTensorMath.cpp"
+#else
+
+using namespace thd;
+using namespace rpc;
+using namespace master;
+
+void THDTensor_(fill)(THDTensor *tensor, real value) {
+  masterCommandChannel->sendMessage(
+    packMessage(
+      Functions::tensorFill,
+      tensor,
+      value
+    ),
+    THDState::s_current_worker
+  );
+}
+
+// TODO implement
+void THDTensor_(zero)(THDTensor *r) {
+  throw std::runtime_error("fill not implemented yet");
+}
+
+void THDTensor_(zeros)(THDTensor *tensor, THLongStorage *size) {
+  THDTensor_(resize)(tensor, size, nullptr);
+  THDTensor_(fill)(tensor, 0);
+}
+
+void THDTensor_(ones)(THDTensor *tensor, THLongStorage *size) {
+  THDTensor_(resize)(tensor, size, nullptr);
+  THDTensor_(fill)(tensor, 1);
+}
+
+ptrdiff_t THDTensor_(numel)(THDTensor *t) {
+  return THDTensor_(nElement)(t);
+}
+
+void THDTensor_(diag)(THDTensor *r_, THDTensor *t, int k) {
+  THArgCheck(THDTensor_(nDimension)(t) == 1 || THDTensor_(nDimension)(t) == 2,
+      1, "matrix or a vector expected");
+
+  if (THDTensor_(nDimension)(t) == 1) {
+    long t_size = THDTensor_(size)(t, 0);
+    long sz = t_size + (k >= 0 ? k : -k);
+
+    THDTensor_(resize2d)(r_, sz, sz);
+    THDTensor_(zero)(r_);
+  } else {
+    long sz;
+    if (k >= 0)
+      sz = THMin(THDTensor_(size)(t, 0), THDTensor_(size)(t, 1)-k);
```

Comment:
I really don't think it's necessary, as it would be redundant, since it is just a [ternary operator](https://github.com/pytorch/pytorch/blob/master/torch/lib/TH/THGeneral.h.in#L118). Same goes for `THArgCheck` etc.